src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests_TestMarkup.cs

```diff
@@ -0,0 +1,404 @@
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Classification;
+
+public partial class SemanticClassifierTests : AbstractCSharpClassifierTests
+{
+    private const string s_testMarkup = """
+
+        static class Test
+        {
+            public static void M([System.Diagnostics.CodeAnalysis.StringSyntax("C#-test")] string code) { }
+        }
+        """ + EmbeddedLanguagesTestConstants.StringSyntaxAttributeCodeCSharp;
+
+    protected async Task TestEmbeddedCSharpAsync(
+       string code,
+       TestHost testHost,
+       params FormattedClassification[] expected)
+    {
+        var allCode = $$"""""
+            class C
+            {
+                void M()
+                {
+                    Test.M(""""
+            {{code}}
+            """");
+                }
+            }
+            """"" + s_testMarkup;
+
+        var start = allCode.IndexOf(code, StringComparison.Ordinal);
+        var length = code.Length;
+        var span = new TextSpan(start, length);
+
+        var actual = await GetClassificationSpansAsync(allCode, span, options: null, testHost);
+
+        // Massage the results a bit so that the TestCode segments don't overlap the non-test-code segments.
+
+        var nonTestCodeSpans = actual.Where(s => s.ClassificationType != ClassificationTypeNames.TestCode).OrderBy((t1, t2) => t1.TextSpan.Start - t2.TextSpan.Start).ToImmutableArray();
+        var testCodeSpans = actual.Where(s => s.ClassificationType == ClassificationTypeNames.TestCode).OrderBy((t1, t2) => t1.TextSpan.Start - t2.TextSpan.Start).ToImmutableArray();
+
+        using var _ = ArrayBuilder<ClassifiedSpan>.GetInstance(out var totalSpans);
+
+        var normalizedNonTestCodeSpans = new NormalizedTextSpanCollection(nonTestCodeSpans.Select(c => c.TextSpan));
+        totalSpans.AddRange(nonTestCodeSpans);
+        foreach (var testCodeSpan in testCodeSpans)
+        {
+            var remainder = NormalizedTextSpanCollection.Difference(new NormalizedTextSpanCollection(testCodeSpan.TextSpan), normalizedNonTestCodeSpans);
+            foreach (var current in remainder)
+            {
+                if (!current.IsEmpty)
+                    totalSpans.Add(new ClassifiedSpan(current, testCodeSpan.ClassificationType));
+            }
+        }
+
+        var actualOrdered = totalSpans.OrderBy(static (t1, t2) => t1.TextSpan.Start - t2.TextSpan.Start).ToImmutableArray();
+        var actualFormatted = actualOrdered.SelectAsArray(a => new FormattedClassification(allCode.Substring(a.TextSpan.Start, a.TextSpan.Length), a.ClassificationType));
+        AssertEx.Equal(expected, actualFormatted);
+    }
+
+    [Theory, CombinatorialData]
+    public async Task TestEmbeddedCSharpMarkup1(TestHost testHost)
+    {
+        await TestEmbeddedCSharpAsync("""
+            class D
+            {
+            }
+            """,
+            testHost,
+            Keyword("class"),
+            TestCode(" "),
```

Comment:
this is embedded C#.  which is why the whitespace gets marked as `TestCode` (incuding that it wwill have a distinct background color).  To make the tests easier to read, we don't put the redundant `TestCode` markers in teh results for spans that we have another embedded classification for.  Otherwise we would have `TestCode("class ")` as well as `Keyword("class")` which just made things a super pain to read and understand.